### ISSUE TITLE
Feature: A way to define plugins

### ISSUE THREAD

**Issue by raphaelkieling** [posva/pinia#191]
A way to create plugin to pinia. I used in my last project the [vuex cache plugin](https://github.com/superwf/vuex-cache) and i would want for a way to create a plugin. Maybe something like:

```ts
// random-project.js
import  { usePlugin } from 'pinia'
import piniaCache from 'pinia-cache'
usePlugin(piniaCache)

// will track both stores
const a = createStore(...)
const b = createStore(...)

// pinia-cache-example.js
export function piniaCache(pinia){
   pinia.subscribe((mutation, state)=> ...)
}
```

This is like a listen store but global to track all state. I dont know if this is the best way but is important have ways to extend the main behavior 😄

**Comment by posva** [comment]
this could be useful!
I'm not sure what the plugin would be receiving, maybe an objects with functions to subscribe to changes and receive the store instance:

```js
function piniaCache({ subscribe, patch }) {
  subscribe((store, otherThings) => {})
  subscribe('storeId', (store) => {})
  patch('storeId', thing)
```

**Comment by raphaelkieling** [comment]
yeah, subscribe with function sounds good. But we would be need different ways do track the data:
- by state
- by actions with hooks

i think that is important to avoid dispatch mutations, actions (listeners) when i only need track a simple state property for example.

```js
function piniaCache({ subscribe, patch }) {
  ...
  subscribeState('storeId', (oldData, newData) => {})
  subscribeBeforeAction('storeId', 'deleteItem', (payload) => {
     return { ...payload, name: 'Kieling' } 
  })
  // will receive name: 'Kieling' into payload
  subscribeAfterAction('storeId', 'deleteItem', (payload) => { })
```

**Comment by posva** [comment]
This has been implemented and documented in #416 and will be available in the next release
